## Importing libraries 

In [4]:
import pandas as pd
import numpy as np
import os

In [6]:
# Defining path
path = r'/Users/tanu/Desktop/Instacart Basket Analysis/02 Data/Prepared data'

In [8]:
# Importing Dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path,'ords_prods_merge_new_columns'))

In [12]:
ords_prods_merge.shape

(32404859, 19)

### Step 2. Aggregated mean of column "order_number" grouped by "department_id"

In [17]:
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3. How the mean for entire dataframe differ from those of the subset with 1000000 rows? 

Mean for the entire data frame is slightly higher than the subset. Department 21 has the highest mean among all the departments in both dataframes followed by department 10. Department 5 and 8 have the least mean of 15.2 and 15.3 respectively .

### Step 4. Create loyality flag for customer based on numbers of order

In [28]:
# Aggregating data with transform function

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/tn/7968l_p11wbcjd89795bwnb40000gn/T/ipykernel_92723/1968777090.py:3: FutureWarning: The provided callable <function max at 0x1041e45e0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [57]:
# checking aggregated data
ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                      2                  8   
1   2539329        1             1                      2                  8   
2   2539329        1             1                      2                  8   
3   2539329        1             1                      2                  8   
4   2539329        1             1                      2                  8   
5   2398795        1             2                      3                  7   
6   2398795        1             2                      3                  7   
7   2398795        1             2                      3                  7   
8   2398795        1             2                      3                  7   
9   2398795        1             2                      3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                     Soda        77              7     9.0   
6                               Pistachios       117             19     3.0   
7                      Original Beef Jerky        23             19     4.4   
8                   Bag of Organic Bananas        24              4    10.3   
9               Aged White Cheddar Popcorn        23             19     4.7   

  Match    price_range_loc   busiest_day     Busiest_days  \
0  both  Mid-range product  Regular busy   Regularly busy   
1  both  Mid-range product  Regular busy   Regularly busy   
2  both  Low-range product  Regular busy   Regularly busy   
3  both  Low-range product  Regular busy   Regularly busy   
4  both  Low-range product  Regular busy   Regularly busy   
5  both  Mid-range product  Regular busy  Least busy days   
6  both  Low-range product  Regular busy  Least busy days   
7  both  Low-range product  Regular busy  Least busy days   
8  both  Mid-range product  Regular busy  Least busy days   
9  both  Low-range product  Regular busy  Least busy days   

  Busiest_period_of_day  max_order  loyalty_flag  
0        Average orders         10  New customer  
1        Average orders         10  New customer  
2        Average orders         10  New customer  
3        Average orders         10  New customer  
4        Average orders         10  New customer  
5        Average orders         10  New customer  
6        Average orders         10  New customer  
7        Average orders         10  New customer  
8        Average orders         10  New customer  
9        Average orders         10  New customer

In [32]:
# creating loyality flag with loc

In [38]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [40]:
# checkung value count for loyalty customer flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

all flags add up to 32404859, total rows in data.

In [50]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                      2                  8   
1   2539329        1             1                      2                  8   
2   2539329        1             1                      2                  8   
3   2539329        1             1                      2                  8   
4   2539329        1             1                      2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  Match    price_range_loc   busiest_day    Busiest_days  \
0  both  Mid-range product  Regular busy  Regularly busy   
1  both  Mid-range product  Regular busy  Regularly busy   
2  both  Low-range product  Regular busy  Regularly busy   
3  both  Low-range product  Regular busy  Regularly busy   
4  both  Low-range product  Regular busy  Regularly busy   

  Busiest_period_of_day  max_order  loyalty_flag  
0        Average orders         10  New customer  
1        Average orders         10  New customer  
2        Average orders         10  New customer  
3        Average orders         10  New customer  
4        Average orders         10  New customer

### Step 5. Find the difference between the spending habits of three types of customer based on loyalty

In [54]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.774457  1.0  25.0
New customer      7.802354  1.0  25.0
Regular customer  7.799319  1.0  25.0

Average expenditure in all three category is almost same,$7.7 - 7.8. And so is the minimun and maximum
We can say, prices of the product purchased by all three type of customers is more or less same.


### Step 6. Create a spending flag for each user based on the average price across all their orders

In [63]:
# Aggregating data with transform function

ords_prods_merge['average_spending']= ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/tn/7968l_p11wbcjd89795bwnb40000gn/T/ipykernel_92723/622719139.py:3: FutureWarning: The provided callable <function mean at 0x1041e4fe0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spending']= ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [69]:
ords_prods_merge[['user_id','average_spending']].head(100)

user_id  average_spending
0         1          6.367797
1         1          6.367797
2         1          6.367797
3         1          6.367797
4         1          6.367797
..      ...               ...
95        2          7.515897
96        2          7.515897
97        2          7.515897
98        2          7.515897
99        2          7.515897

[100 rows x 2 columns]

In [71]:
# Creating the spending label
ords_prods_merge.loc[ords_prods_merge['average_spending'] < 10 , 'spending_label'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['average_spending'] >= 10, 'spending_label'] = 'High spender'

In [73]:
# Counting the values 
ords_prods_merge['spending_label'].value_counts(dropna = False)

spending_label
Low spender     32284898
High spender      119961
Name: count, dtype: int64

99% of customers are in low spending flag.

In [81]:
# checking relevant columns
ords_prods_merge[['user_id','average_spending','spending_label']].head(100)

user_id  average_spending spending_label
0         1          6.367797    Low spender
1         1          6.367797    Low spender
2         1          6.367797    Low spender
3         1          6.367797    Low spender
4         1          6.367797    Low spender
..      ...               ...            ...
95        2          7.515897    Low spender
96        2          7.515897    Low spender
97        2          7.515897    Low spender
98        2          7.515897    Low spender
99        2          7.515897    Low spender

[100 rows x 3 columns]

### Step 7. create an order frequency flag

In [86]:
# aggregating data with transform function
ords_prods_merge['days_btwn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/tn/7968l_p11wbcjd89795bwnb40000gn/T/ipykernel_92723/667489865.py:2: FutureWarning: The provided callable <function median at 0x104320680> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['days_btwn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [90]:
ords_prods_merge[['user_id','days_since_prior_order','days_btwn_orders']].head()

user_id  days_since_prior_order  days_btwn_orders
0        1                     7.0              20.0
1        1                     7.0              20.0
2        1                     7.0              20.0
3        1                     7.0              20.0
4        1                     7.0              20.0

In [92]:
# Creating 'Order_frequency' flag
ords_prods_merge.loc[ords_prods_merge['days_btwn_orders'] > 20, 'order_frequency'] = 'Non frequent customer'
ords_prods_merge.loc[(ords_prods_merge['days_btwn_orders'] <= 20) & (ords_prods_merge['days_btwn_orders'] >10), 'order_frequency'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['days_btwn_orders'] <= 10, 'order_frequency'] = 'Frequent customer'

In [94]:
# counting values
ords_prods_merge['order_frequency'].value_counts(dropna= False)

order_frequency
Frequent customer        22790523
Regular customer          6927608
Non frequent customer     2686728
Name: count, dtype: int64

70 % customers order frequently,
21% orders regularly,
and 8 % customers orders rarely

In [118]:
ords_prods_merge.head(100)

order_id  user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0    2539329        1             1                      2                  8   
1    2539329        1             1                      2                  8   
2    2539329        1             1                      2                  8   
3    2539329        1             1                      2                  8   
4    2539329        1             1                      2                  8   
..       ...      ...           ...                    ...                ...   
95    738281        2             4                      2                 10   
96   1673511        2             5                      3                 11   
97   1673511        2             5                      3                 11   
98   1673511        2             5                      3                 11   
99   1673511        2             5                      3                 11   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      7.0         196                  1          0   
1                      7.0       14084                  2          0   
2                      7.0       12427                  3          0   
3                      7.0       26088                  4          0   
4                      7.0       26405                  5          0   
..                     ...         ...                ...        ...   
95                     8.0       21150                 13          0   
96                     8.0       47144                  1          0   
97                     8.0        5322                  2          0   
98                     8.0       17224                  3          0   
99                     8.0       38656                  4          0   

                                         product_name  ...    price_range_loc  \
0                                                Soda  ...  Mid-range product   
1             Organic Unsweetened Vanilla Almond Milk  ...  Mid-range product   
2                                 Original Beef Jerky  ...  Low-range product   
3                          Aged White Cheddar Popcorn  ...  Low-range product   
4                    XL Pick-A-Size Paper Towel Rolls  ...  Low-range product   
..                                                ...  ...                ...   
95                            Fire Grilled Steak Bowl  ...  Mid-range product   
96     Unsweetened Original Almond Breeze Almond Milk  ...  Mid-range product   
97  Gluten Free Dark Chocolate Chunk Chewy with a ...  ...  Low-range product   
98                   Oats & Honey Gluten Free Granola  ...  Low-range product   
99  Gluten Free Peanut Butter Dark Chocolate Chewy...  ...  Low-range product   

     busiest_day     Busiest_days Busiest_period_of_day max_order  \
0   Regular busy   Regularly busy        Average orders        10   
1   Regular busy   Regularly busy        Average orders        10   
2   Regular busy   Regularly busy        Average orders        10   
3   Regular busy   Regularly busy        Average orders        10   
4   Regular busy   Regularly busy        Average orders        10   
..           ...              ...                   ...       ...   
95  Regular busy   Regularly busy           Most orders        14   
96  Regular busy  Least busy days           Most orders        14   
97  Regular busy  Least busy days           Most orders        14   
98  Regular busy  Least busy days           Most orders        14   
99  Regular busy  Least busy days           Most orders        14   

        loyalty_flag average_spending spending_label  days_btwn_orders  \
0       New customer         6.367797    Low spender              20.0   
1       New customer         6.367797    Low spender              20.0   
2       New customer         6.367797    Low spender              20.0   
3       New customer         6.367797    Low spender              20

### Step 8. Exporting the data


In [120]:
ords_prods_merge.to_pickle(os.path.join(path,'ords_prods_merged_grouped.pkl'))